students/terrance_jones/lesson5/mailroom_part3.py

```diff
@@ -0,0 +1,169 @@
+#list of donors
+donor_list = {'Carlos Santos':[25,50,100], 'Esperanza Gomez': [10,20,30], 'Paul Jackson':[5,10,15], 'Karl Black':[100,200,300], 'Charles Exx': [15,30,45]}
+
+
+"""
+FUNCTIONS NEEDED FOR SCRIPT TO RUN
+"""
+def get_name():
+    """get donor name from user"""
+    name = input("Enter the full name of the donor.")
+    return name
+
+def show_donor_list():
+    """Create list of donor names only. Display list of all donors"""
+    for key in donor_list.keys():
+        print(key)
+    
+
+def find_donor(x):
+    for key in donor_list.keys():
+        if key == x:
+            return True
+        
+
+def get_amount():
+    """Gets donation amount from user."""
+    while True:
+        try:
+            amount = input("Enter the donation amount:")
+            amount = float(amount)
+        except ValueError:
+            print("Donation amount must be a numeric value")
+        
+        else:
+            return amount
+
+def add_to_existing_donor(d, a):
+    """adds donation amount to donor in list. d is donor name. a is amount of donation"""
+    donor_list[d].append(a)
+
+def create_new_donor(d, a):
+    donor_list[d] = [a]
+
+def print_rows():
+    for c in donor_list:
+        name = c
+        num_gifts = len(donor_list[c])
+        total = sum(donor_list[c])
+        average = total / num_gifts
+        average = round(average,2)
+
+        
+        
+        print("{:<25} ${:^20}  {:^15} ${:>20}".format(name,total,num_gifts,average))
+
+
+def letter(donor):
+    output = {'name': donor}
+    all_donations = donor_list[donor]
+    output = {'name': donor, 'total': sum(donor_list[donor])}
+   
+    output['last_donation']= all_donations[len(all_donations)-1]
+    letter = '''Dear {name},\n
+    \tThank you for your most recent donation of ${last_donation:,.2f}
+    \tYour total support of {total:,.2f} has been used to assist many people.\n
+    Thank you,\n
+    -The Mailroom'''.format(**output)
+
+    return(letter)
+
+
+
+#thank you function gets a donors name and amount of donation. then prints out a thank you email
+def thankyou():
+    donor_name = get_name()
+    if donor_name == "list":
+        show_donor_list
+        donor_name =get_name()
+
+    if find_donor(donor_name) is True: 
+        donation_amount = get_amount()
+        add_to_existing_donor(donor_name, donation_amount)
+        
+        dic = {'name': donor_name , 'amount_donated': donation_amount}
+        print("Dear {name}, \n Thank you for your {amount_donated:,.2f} donation. \nSincerely, Mailroom.".format(**dic))
+        
+    else:
+        
+        donation_amount = get_amount()
+        create_new_donor(donor_name, donation_amount)
+        dic = {'name': donor_name , 'amount_donated': donation_amount}
+        print("Dear {name}, \n Thank you for your {amount_donated:,.2f} donation. \nSincerely, Mailroom.".format(**dic))
```

Comment:
lines 81-85 and 89-92 are identical except add_to_existing_donor/create_new_donor function calls - you generally want to consolidate duplicate code, how about:
```
donation_amount = get_amount()
add_or_update_donor(donor_name, donor_amount)  # let this function do the appropriate work
dic = {'name': donor_name , 'amount_donated': donation_amount}
print("Dear {name}, \n Thank you for your {amount_donated:,.2f} donation. \nSincerely, Mailroom.".format(**dic))
```